trieste/acquisition/function/function.py

```diff
@@ -583,6 +583,217 @@ def _update_expected_improvement_fn(
             self._expected_improvement_fn.update(eta)  # type: ignore
 
 
+class MonteCarloExpectedImprovement(SingleModelAcquisitionBuilder[ProbabilisticModel]):
+    """
+    Builder for a Monte Carlo-based expected improvement function for use with a model without
+    analytical expected improvement (e.g. a deep GP). The "best" value is taken to be
+    the minimum of the posterior mean at observed points.
+    """
+
+    def __init__(self, sample_size: int):
+        """
+        :param sample_size: The number of samples for each batch of points.
+        :raise tf.errors.InvalidArgumentError: If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        super().__init__()
+
+        self._sample_size = sample_size
+
+    def __repr__(self) -> str:
+        """"""
+        return f"MonteCarloExpectedImprovement({self._sample_size!r})"
+
+    def prepare_acquisition_function(
+        self,
+        model: ProbabilisticModel,
+        dataset: Optional[Dataset] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param model: The model over the specified ``dataset``. Must have output dimension [1].
+        :param dataset: The data from the observer. Cannot be empty.
+        :return: The estimated *expected improvement* acquisition function.
+        :raise ValueError (or InvalidArgumentError): If ``dataset`` is not populated, ``model``
+            does not have an output dimension of [1] or does not have a ``reparam_sample`` method.
+        """
+        get_sampler = getattr(model, "reparam_sampler", None)
+        if not callable(get_sampler):
+            raise ValueError(
+                "The model must have a `reparam_sampler` method, which takes `num_samples` and "
+                "which returns a reparametrization `ReparametrizationSampler` to create smooth "
+                "samples at query points."
+            )
+
+        sampler = get_sampler(self._sample_size)
+
+        if not isinstance(sampler, ReparametrizationSampler):
+            raise ValueError(
+                "The sampler returned by the model's `reparam_sampler` method must be an instance "
+                "of a `trieste.models.interfaces.ReparametrizationSampler`."
+            )
+
+        if not isinstance(dataset, Dataset):
+            raise ValueError(
+                f"`dataset` must be a Dataset for {self.__repr__()}, received" f"{type(dataset)}"
+            )
+
+        tf.debugging.assert_positive(len(dataset))
+
+        samples_at_query_points = sampler.sample(dataset.query_points)
+        mean = tf.reduce_mean(samples_at_query_points, axis=0)
+
+        tf.debugging.assert_shapes(
+            [(mean, ["_", 1])], message="Expected model with output dimension [1]."
+        )
+
+        eta = tf.reduce_min(mean, axis=0)
+
+        return monte_carlo_expected_improvement(sampler, eta)
+
+
+class monte_carlo_expected_improvement(AcquisitionFunctionClass):
```

Comment:
when Henry refactored samplers 2 months ago he changed `batch_monte_carlo_expected_improvement` design, probably you should emulate it here with update functions, resetting the sampler etc